### ISSUE TITLE
UIKit calls periodically made off the main thread crashes

### ISSUE THREAD

**Issue by jeffremer** [zendesk/zendesk_sdk_ios#396]
This may either be very related or the same as #390. The Zendesk SDK is currently the leading cause of crashes in our app (Strava). It appears that Zendesk is calling UIKit off the main thread, trying to render a web view in the background. Below are some examples of the main thread sanitizer catching an instance of this and I've also attached an example stack trace we've gathered from Crashlytics.

```
=================================================================
Main Thread Checker: UI API called on a background thread: -[UIView setBackgroundColor:]
PID: 2291, TID: 1141372, Thread name: (none), Queue name: com.apple.root.default-qos, QoS: 21
Backtrace:
4   Apptimize                           0x000000010d94876c __ABTSwizzleDynamicSelectorProperty_block_invoke + 576
5   ZendeskSDK                          0x000000010f787a24 -[ZDKArticleView renderArticle:withHelpCenterSettings:] + 136
6   ZendeskSDK                          0x000000010f799be4 __59-[ZDKArticleViewController commonViewModelInit:andConfigs:]_block_invoke_2 + 192
7   ZendeskProviderSDK                  0x000000010f65bea8 __62-[ZDKHelpCenterProvider getArticleWithId:locale:withCallback:]_block_invoke.352 + 80
8   ZendeskProviderSDK                  0x000000010f66f8dc __66-[ZDKAPIDispatcher executeRequest:onSuccess:onError:requiresAuth:]_block_invoke + 116
9   ZendeskProviderSDK                  0x000000010f683e64 _T0So11URLResponseCSgSo6NSDataCSgSo7NSErrorCSgIeyByyy_AC10Foundation4DataVSgs5Error_pSgIegxxx_TR + 108
10  ZendeskProviderSDK                  0x000000010f6865e0 _T018ZendeskProviderSDK17DispatcherAdapterC14performRequesty0a4CoreC00A0C7zendesk_10Foundation10URLRequestV10urlRequsetSb12requiresAuthySo11URLResponseCSg_AI4DataVSgs5Error_pSgtc17completionHandlertFZyAS_ApUtcfU_TA + 96
11  ZendeskCoreSDK                      0x000000010fe88b84 _T014ZendeskCoreSDK13NetworkClientC7executeyAA15ProviderServiceV7service_So13DispatchQueueC2ony10Foundation4DataVSg_So11URLResponseCSgs5Error_pSgtc17completionHandlertF05_T014ab41SDK22DefaultRequestExecutorC7performyAA15gh28V08providerI0_y10Foundation4n8VSg_So11o5CSgs5P36_pSgtc8callbacktFZyycfU_yAK_AnPtcfU_AnqSIegxxx_Tf1nncn_nTf4ngnn_n + 608
12  ZendeskCoreSDK                      0x000000010fe911d8 _T014ZendeskCoreSDK22DefaultRequestExecutorC7performyAA15ProviderServiceV08providerI0_y10Foundation4DataVSg_So11URLResponseCSgs5Error_pSgtc8callbacktFZyycfU_ + 148
13  ZendeskCoreSDK                      0x000000010fe912b4 _T0Ieg_IeyB_TR + 36
14  libdispatch.dylib                   0x0000000110ac51dc _dispatch_call_block_and_release + 24
15  libdispatch.dylib                   0x0000000110ac519c _dispatch_client_callout + 16
16  libdispatch.dylib                   0x0000000110ad1200 _dispatch_queue_override_invoke + 968
17  libdispatch.dylib                   0x0000000110ad67c8 _dispatch_root_queue_drain + 604
18  libdispatch.dylib                   0x0000000110ad6500 _dispatch_worker_thread3 + 136
19  libsystem_pthread.dylib             0x0000000180d13fac _pthread_wqthread + 1176
20  libsystem_pthread.dylib             0x0000000180d13b08 start_wqthread + 4
2018-06-06 06:44:19.227342-0700 Strava[2291:1141372] [reports] Main Thread Checker: UI API called on a background thread: -[UIView setBackgroundColor:]
PID: 2291, TID: 1141372, Thread name: (none), Queue name: com.apple.root.default-qos, QoS: 21
Backtrace:
4   Apptimize                           0x000000010d94876c __ABTSwizzleDynamicSelectorProperty_block_invoke + 576
5   ZendeskSDK                          0x000000010f787a24 -[ZDKArticleView renderArticle:withHelpCenterSettings:] + 136
6   ZendeskSDK                          0x000000010f799be4 __59-[ZDKArticleViewController commonViewModelInit:andConfigs:]_block_invoke_2 + 192
7   ZendeskProviderSDK                  0x000000010f65bea8 __62-[ZDKHelpCenterProvider getArticleWithId:locale:withCallback:]_block_invoke.352 + 80
8   ZendeskProviderSDK                  0x000000010f66f8dc __66-[ZDKAPIDispatcher executeRequest:onSuccess:onError:requiresAuth:]_block_invoke + 116
9   ZendeskProviderSDK                  0x000000010f683e64 _T0So11URLResponseCSgSo6NSDataCSgSo7NSErrorCSgIeyByyy_AC10Foundation4DataVSgs5Error_pSgIegxxx_TR + 108
10  ZendeskProviderSDK                  0x000000010f6865e0 _T018ZendeskProviderSDK17DispatcherAdapterC14performRequesty0a4CoreC00A0C7zendesk_10Foundation10URLRequestV10urlRequsetSb12requiresAuthySo11URLResponseCSg_AI4DataVSgs5Error_pSgtc17completionHandlertFZyAS_ApUtcfU_TA + 96
11  ZendeskCoreSDK                      0x000000010fe88b84 _T014ZendeskCoreSDK13NetworkClientC7executeyAA15ProviderServiceV7service_So13DispatchQueueC2ony10Foundation4DataVSg_So11URLResponseCSgs5Error_pSgtc17completionHandlertF05_T014ab41SDK22DefaultRequestExecutorC7performyAA15gh28V08providerI0_y10Foundation4n8VSg_So11o5CSgs5P36_pSgtc8callbacktFZyycfU_yAK_AnPtcfU_AnqSIegxxx_Tf1nncn_nTf4ngnn_n + 608
12  ZendeskCoreSDK                      0x000000010fe911d8 _T014ZendeskCoreSDK22DefaultRequestExecutorC7performyAA15ProviderServiceV08providerI0_y10Foundation4DataVSg_So11URLResponseCSgs5Error_pSgtc8callbacktFZyycfU_ + 148
13  ZendeskCoreSDK                      0x000000010fe912b4 _T0Ieg_IeyB_TR + 36
14  libdispatch.dylib                   0x0000000110ac51dc _dispatch_call_block_and_release + 24
15  libdispatch.dylib                   0x0000000110ac519c _dispatch_client_callout + 16
16  libdispatch.dylib                   0x0000000110ad1200 _dispatch_queue_override_invoke + 968
17  libdispatch.dylib                   0x0000000110ad67c8 _dispatch_root_queue_drain + 604
18  libdispatch.dylib                   0x0000000110ad6500 _dispatch_worker_thread3 + 136
19  libsystem_pthread.dylib             0x0000000180d13fac _pthread_wqthread + 1176
20  libsystem_pthread.dylib             0x0000000180d13b08 start_wqthread + 4
2018-06-06 06:44:19.614135-0700 Strava[2291:1141372] void _WebThreadLock(), 0x112e9c000: Tried to obtain the web lock from a thread other than the main thread or the web thread. This may be a result of calling to UIKit from a secondary thread. Crashing now...
```

```
2018-06-06 06:53:42.908217-0700 Strava[2308:1147048] This application is modifying the autolayout engine from a background thread after the engine was accessed from the main thread. This can lead to engine corruption and weird crashes.
 Stack:(
	0   Foundation                          0x0000000181c48424 <redacted> + 96
	1   Foundation                          0x0000000181a3d4ac <redacted> + 36
	2   UIKit                               0x000000018acd3c40 <redacted> + 780
	3   UIKit                               0x000000018acd2b60 <redacted> + 1732
	4   UIKit                               0x000000018afdf918 <redacted> + 416
	5   UIKit                               0x000000018ad8f580 <redacted> + 212
	6   UIKit                               0x000000018ad92860 <redacted> + 328
	7   UIKit                               0x000000018aceca88 <redacted> + 172
	8   UIKit                               0x000000018acec5a0 <redacted> + 1376
	9   UIKit                               0x000000018ad8be00 <redacted> + 92
	10  UIKit                               0x000000018afbe184 <redacted> + 156
	11  Foundation                          0x0000000181a3d500 <redacted> + 120
	12  UIKit                               0x000000018acd3c40 <redacted> + 780
	13  UIKit                               0x000000018acd2b60 <redacted> + 1732
	14  ZendeskSDK                          0x0000000112f57b70 -[ZDKArticleView renderArticle:withHelpCenterSettings:] + 468
	15  ZendeskSDK                          0x0000000112f69be4 __59-[ZDKArticleViewController commonViewModelInit:andConfigs:]_block_invoke_2 + 192
	16  ZendeskProviderSDK                  0x0000000112e2bea8 __62-[ZDKHelpCenterProvider getArticleWithId:locale:withCallback:]_block_invoke.352 + 80
	17  ZendeskProviderSDK                  0x0000000112e3f8dc __66-[ZDKAPIDispatcher executeRequest:onSuccess:onError:requiresAuth:]_block_invoke + 116
	18  ZendeskProviderSDK                  0x0000000112e53e64 _T0So11URLResponseCSgSo6NSDataCSgSo7NSErrorCSgIeyByyy_AC10Foundation4DataVSgs5Error_pSgIegxxx_TR + 108
	19  ZendeskProviderSDK                  0x0000000112e565e0 _T018ZendeskProviderSDK17DispatcherAdapterC14performRequesty0a4CoreC00A0C7zendesk_10Foundation10URLRequestV10urlRequsetSb12requiresAuthySo11URLResponseCSg_AI4DataVSgs5Error_pSgtc17completionHandlertFZyAS_ApUtcfU_TA + 96
	20  ZendeskCoreSDK                      0x0000000113658b84 _T014ZendeskCoreSDK13NetworkClientC7executeyAA15ProviderServiceV7service_So13DispatchQueueC2ony10Foundation4DataVSg_So11URLResponseCSgs5Error_pSgtc17completionHandlertF05_T014ab41SDK22DefaultRequestExecutorC7performyAA15gh28V08providerI0_y10Foundation4n8VSg_So11o5CSgs5P36_pSgtc8callbacktFZyycfU_yAK_AnPtcfU_AnqSIegxxx_Tf1nncn_nTf4ngnn_n + 608
	21  ZendeskCoreSDK                      0x00000001136611d8 _T014ZendeskCoreSDK22DefaultRequestExecutorC7performyAA15ProviderServiceV08providerI0_y10Foundation4DataVSg_So11URLResponseCSgs5Error_pSgtc8callbacktFZyycfU_ + 148
	22  ZendeskCoreSDK                      0x00000001136612b4 _T0Ieg_IeyB_TR + 36
	23  libdispatch.dylib                   0x00000001142b51dc _dispatch_call_block_and_release + 24
	24  libdispatch.dylib                   0x00000001142b519c _dispatch_client_callout + 16
	25  libdispatch.dylib                   0x00000001142c1200 _dispatch_queue_override_invoke + 968
	26  libdispatch.dylib                   0x00000001142c67c8 _dispatch_root_queue_drain + 604
	27  libdispatch.dylib                   0x00000001142c6500 _dispatch_worker_thread3 + 136
	28  libsystem_pthread.dylib             0x0000000180d13fac _pthread_wqthread + 1176
	29  libsystem_pthread.dylib             0x0000000180d13b08 start_wqthread + 4
)
2018-06-06 06:53:42.913202-0700 Strava[2308:1147048] void _WebThreadLock(), 0x120c34000: Tried to obtain the web lock from a thread other than the main thread or the web thread. This may be a result of calling to UIKit from a secondary thread. Crashing now...
```

![screen shot 2018-06-06 at 6 48 37 am](https://user-images.githubusercontent.com/99528/41042107-a90d06a4-6955-11e8-9831-0527a0216a70.png)

- SDK version: 2.0.3
- Xcode version: 9.3
- iOS or Simulator: Device
- iOS Version: 11.3.0-11.4.1
- Swift version: 4.1

Steps to Reproduce:

1. View a single article, quickly, for the first time after installing and launching the app

[com.strava.stravaride_issue_2455_crash_7f176c8754784bc0a2764f0fbd12ae76_DNE_0_v2.txt](https://github.com/zendesk/zendesk_sdk_ios/files/2076526/com.strava.stravaride_issue_2455_crash_7f176c8754784bc0a2764f0fbd12ae76_DNE_0_v2.txt)

**Comment by KillianSmith0** [comment]
Hi @jeffremer,

Thanks for the heads up, I've also come across this crash while debugging the previous issue. I've prioritised these problems and plan to have a solution as soon as possible, which will be resolved in the next patch. You'll be alerted when this occurs
Sorry for the inconvenience

**Comment by dhsu210** [comment]
Hi @KillianSmith0 I've also been experiencing the same issue. Not sure if it helps to look at my crash logs, which I've referenced in the [developer community site](https://develop.zendesk.com/hc/en-us/community/posts/360002892708-iOS-v2-article-view-crashing-messages-not-going-through-when-testing-on-iPhone). 

SDK version: 2.0.3
Xcode version: 9.4 beta
iOS or Simulator: iPhone 8 Plus
iOS Version:
Swift version: 3

**Comment by dhsu210** [comment]
@jeffremer Strange that this crash isn't occurring periodically for me, but pretty consistently 100% of the time.

**Comment by jeffremer** [comment]
> @jeffremer Strange that this crash isn't occurring periodically for me, but pretty consistently 100% of the time.

I can reproduce it consistently from a fresh install and if I view the article quickly - however viewing the article after it has been loaded once (and maybe cached?) doesn't crash.

**Comment by dhsu210** [comment]
@jeffremer while we're waiting for the latest release SDK, I was wondering -- are you presenting the article view directly on top of the VC? What I'm currently doing is adding it as a childVC to a navigation controller, and then presenting that navigation controller. Not sure if this might have something to do with the crashing.

**Comment by jeffremer** [comment]
Not exactly @dhsu210 - this is happening in our app which has an abstraction around it so that code isn't really helpful to look at, but it boils down to the same code in #390 - I attached a sample project in that ticket as well.

**Comment by dhsu210** [comment]
@jeffremer It doesn't seem to be crashing for me anymore, not sure if they fixed something remotely on their end. just FYI

**Comment by dhsu210** [comment]
@KillianSmith0 I noticed this crash only occurs for me when I am signed in with the same email in our app as Zendesk admin/agent email. If I am logged in my app with a regular customer's email, then the articles load without crashing. Hope this helps!

**Comment by brendan-fahy** [comment]
Hey @dhsu210, currently the back-end rejects admin/agent emails, we get 403s if you try to use them. This has come up a few times and there are ongoing conversations about changing that behaviour, but for the moment, that's how it is. That being said, we shouldn't be crashing just because we got a 403. That sounds bad. Is it specifically viewing articles that crashes for you in this case?

**Comment by dhsu210** [comment]
@brendan-fahy Yes, when using an admin/agent email it causes a crash when loading an article from the help center. Does not crash when I use a non admin/agent email.

**Comment by KillianSmith0** [comment]
@dhsu210 The crash in relation to the Article VC is caused by UI stuff being called on a background thread, judging by your previous comments it sounds like your embedding the VC in one too many views which is causing an overload, resulting in an earlier crash than usual. The fix in the patch should resolve your issue, irregardless you should probably take a closer look into how you're presenting an article

**Comment by jeffremer** [comment]
@KillianSmith0 any update/ETA? This remains our largest crash, by an order of magnitude - it's affecting thousands of users.

**Comment by KillianSmith0** [comment]
@jeffremer - Unfortunately I'm unable to promise you a specific ETA, but this fix will be available in our next release for iOS. I'll keep you posted as soon as it goes out.

**Comment by KillianSmith0** [comment]
@jeffremer Hi Jeff, we have just release the update for this fix on v2.0.4. This update includes a fix for both your open issues. Thanks for your patience

**Comment by jeffremer** [comment]
Thanks @KillianSmith0, we've updated to 2.0.5. Much appreciated. 👍
